### DISCUSSION TITLE
How to access the current user session data through api routes using NextJS

### DISCUSSION THREAD

**Discussion by pablolucio97** [nextauthjs/next-auth#3386] [Help] [answered]
### Question 💬

0

I'm trying to access data from the current user session after user authentication in a file inside the Api routes from NextJS, but in the console.log(session.user.email) returns null even after the authentication. Does someone know how to do it?

```
import { NextApiRequest, NextApiResponse } from "next";
import connectDb from "../../services/mongo";
import connectToDatabase from "../../services/mongo";
import {getSession} from 'next-auth/client'

connectToDatabase();

export default async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  const { db } = await connectDb();

  switch (req.method) {
    case "GET":
      try {
        const session = await getSession()
        console.log(session.user.email)
        const favorites = await db.collection("users").findOne({  email: session.user.email})
        res.status(200).json({ favorites });
      } catch (error) {
        console.log(error);
      }
      break;
    default:
  }
};
```

### How to reproduce ☕️

1) User do signIn in the application using some provider GitHub, Google, or Discord.
2) User access /dashboard page and see your all favorite games based on user id.

The problem is that I can't get the current user from session.user.email  in api/users route to list all favorite user games correctly.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
you have to pass `req`

https://next-auth.js.org/getting-started/client#server-side-example

**Reply by meAnurag** [reply]
Hey, How can I achieve the same in App Router's Route Handlers ?

**Comment by 1finedev** [comment, marked as answer]
```const session = await getSession({req})```

**Reply by Simonbrl** [reply]
I tried but I get this error : 
`[next-auth][error][CLIENT_FETCH_ERROR]
https://next-auth.js.org/errors#client_fetch_error "[object Object]" is not valid JSON {
  error: {
    message: '"[object Object]" is not valid JSON',
    stack: 'SyntaxError: "[object Object]" is not valid JSON\n' +
      '    at JSON.parse (<anonymous>)\n' +
      '    at parseJSONFromBytes (node:internal/deps/undici/undici:6680:19)\n' +
      '    at successSteps (node:internal/deps/undici/undici:6654:27)\n' +
      '    at node:internal/deps/undici/undici:1254:60\n' +
      '    at node:internal/process/task_queues:140:7\n' +
      '    at AsyncResource.runInAsyncScope (node:async_hooks:206:9)\n' +
      '    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)\n' +
      '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)',
    name: 'SyntaxError'
  },
  url: 'http://localhost:3000/api/auth/session',
  message: '"[object Object]" is not valid JSON'
}
null`

My code in app/api/users/routes.ts
`import { NextApiRequest } from "next";
import { getSession } from "next-auth/react";

export const GET = async (req: NextApiRequest) => {
    const session = await getSession({ req });
    console.log(session);
}`

Someone would have any idea ?

**Reply by piclez** [reply]
For latest version:

```
const session = await auth();
import { auth } from "@/auth"
```

**Comment by pablolucio97** [comment]
Thank you very much! It has worked fine! If desired, consider this topic as closed.

**Comment by RamK777-stack** [comment]
In post method I am getting session is null but in GET request, able get session. how ?

**Reply by tomryanx** [reply]
Same here... troubling

**Reply by adamjakab** [reply]
With 
`const session = await getSession({ req })`
the result will be `null ` when the request method ie either **POST** or **PUT** (I have not tested others) but it will correctly return the Session object when using **GET**. 
NB.: Cookies (next-auth.csrf-token and next-auth.session-token)are correctly sent with the requests. 
When requesting the session, backend will log:
```
[next-auth][error][CLIENT_FETCH_ERROR]
https://next-auth.js.org/errors#client_fetch_error undefined {
  error: {},
  url: 'http://localhost:3000/api/auth/session',
  message: undefined
}
```

Any suggestions?

**Reply by adamjakab** [reply]
Apologies! RTFM ;) here: https://next-auth.js.org/tutorials/securing-pages-and-api-routes#using-getserversession
It says explicitly: "_You can protect API routes using the getServerSession() method._"

The below code will work independently of the request method:

```
import type { NextApiRequest, NextApiResponse } from "next"
import { getServerSession } from "next-auth"
import { authOptions } from "./auth/[...nextauth]"
/**
 * Store user preferences
 * @file /api/preferences.ts
 * @param req
 * @param res
 */
const PreferencesStorage = async (req: NextApiRequest, res: NextApiResponse) => {
    const session = await getServerSession(req, res, authOptions)
    console.log("Session: ", session)
    
    res.status(200).json({"msg":"ok"})
}

export default PreferencesStorage

```

**Reply by DavidForDev** [reply]
hi everyone, I use "connectionParams" in the nextjs project to pass user session to nodejs, like that:

next.js: 

```
const wsLink =
  typeof window !== "undefined"
    ? new GraphQLWsLink(
        createClient({
          url: "ws://localhost:5000/subscriptions",
          connectionParams: async () => ({
            session: await getSession(),
          }),
        })
      )
    : null;

const httpLink = new HttpLink({
  uri: `http://localhost:5000/`,
  credentials: "include",
});
  // ....Other code

```

and node.js: 
```
// ============= apollo server ============== \\
  const server = new ApolloServer<any>({
    schema: schem,
    csrfPrevention: true,
    plugins: [
      ApolloServerPluginDrainHttpServer({ httpServer }),
      ApolloServerPluginLandingPageLocalDefault({
        embed: true,
        includeCookies: true,
      }),
      // Proper shutdown for the HTTP server.
      {
        async serverWillStart() {
          return {
            async drainServer() {
              await serverCleanup.dispose();
            },
          };
        },
      },
    ],
  });

  await server.start();

  app.use(
    "/",
    cors<cors.CorsRequest>({ credentials: true, origin: "*" }),
    json(),
    expressMiddleware(server, {
      context: async ({ req }): Promise<GraphqlContext> => {
        const session = await getSession({ req });

        console.log(session);

        return { pubSub, prisma, session: session as Session };
      },
    })
  );
```


but still I have same problem, I get such error: 

[next-auth][error][CLIENT_FETCH_ERROR] 
https://next-auth.js.org/errors#client_fetch_error "[object Object]" is not valid JSON {
  error: {
    message: '"[object Object]" is not valid JSON',
    stack: 'SyntaxError: "[object Object]" is not valid JSON\n' +
      '    at JSON.parse (<anonymous>)\n' +
      '    at parseJSONFromBytes (node:internal/deps/undici/undici:4763:19)\n' +
      '    at successSteps (node:internal/deps/undici/undici:4734:27)\n' +
      '    at fullyReadBody (node:internal/deps/undici/undici:1447:9)\n' +
      '    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
      '    at async specConsumeBody (node:internal/deps/undici/undici:4743:7)',
    name: 'SyntaxError'
  },
  url: 'http://localhost:3000/api/auth/session',
  message: '"[object Object]" is not valid JSON'
}
null


if you have any solutions please let me know, I tried a lot of ways to solve it but nothing works

**Reply by SushWar** [reply]
Hi David,

I also faced this issue but there is a way around, Hope this will help you and others!! 

```
   import * as jwt from "next-auth/jwt"

  app.use(
    "/",
    cors<cors.CorsRequest>({ credentials: true, origin: "*" }),
    json(),
    expressMiddleware(server, {
      context: async ({ req }): Promise<GraphqlContext> => {
        const headCookies = req?.headers.cookie?.split(";")
        const sessiontoken = headCookies?.at(2)?.split("=")
       
         try {
          const decode = await jwt.decode({
            token: sessiontoken?.at(1),
            secret: String(process.env.NEXTAUTH_SECRET),
          })

          if (!decode) {
            console.log("Decode is null")
          }
          console.log(decode)
          return { pubSub, prisma, session: decode as Session };
        } catch (error) {
          console.log("Error ----------> ", error)
          return {error}
        }

        
      },
    })
  );
```

**Comment by danikane** [comment]
`getToken` worked for me inside an api route with the pages router using NextJS 13.
https://next-auth.js.org/tutorials/securing-pages-and-api-routes#using-gettoken
Looking forward to using `AuthJS` :)
